### PR TITLE
Bug 1909502: pkg/operator: tolerate removal of etcd records from proxy config

### PR DESCRIPTION
**Pull request by hexfusion**
Since 4.4 etcd no longer has a DNS dependency. Currently, we are still passing etcd records to the proxy config which should be removed.

In order to remove the deprecated etcd records from the proxy config, we must tolerate the old etcd records.
This PR allows the toleration of this divergence providing the user a chance to update the NoProxy config and in the meantime allow CI to pass for the below PRs.

blocks:
- https://github.com/openshift/installer/pull/4518
- https://github.com/openshift/cluster-network-operator/pull/930

### CODE REVIEW

**Comment by hexfusion** [comment]
cluster bot proxy test
https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws/1346306484308283392

**Comment by hexfusion** [comment]
From proxy test cluster
https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/release-openshift-origin-installer-launch-aws/1346306484308283392

```
/etc/machine-config-daemon/currentconfig
NO_PROXY=.cluster.local,.svc,.us-west-1.compute.internal,10.0.0.0/16,10.128.0.0/14,127.0.0.1,169.254.169.254,172.30.0.0/16,api-int.ci-ln-w076dpt-d5d6b.origin-ci-int-aws.dev.rhcloud.com,localhost\n
```

```
/etc/mcs-machine-config-content.json:338:                                "contents": "[Unit]\n[Service]\nEnvironment=HTTP_PROXY=http://ewolinet:5f6ccbbbafc66013d012839921ada773@35.196.128.173:3128/\nEnvironment=HTTPS_PROXY=http://ewolinet:5f6ccbbbafc66013d012839921ada773@35.196.128.173:3128/\nEnvironment=NO_PROXY=.cluster.local,.svc,.us-west-1.compute.internal,10.0.0.0/16,10.128.0.0/14,127.0.0.1,169.254.169.254,172.30.0.0/16,api-int.ci-ln-w076dpt-d5d6b.origin-ci-int-aws.dev.rhcloud.com,localhost\n",
```

```
 $ oc get proxy -o json | jq '.items[0].status.noProxy'
".cluster.local,.svc,.us-west-1.compute.internal,10.0.0.0/16,10.128.0.0/14,127.0.0.1,169.254.169.254,172.30.0.0/16,api-int.ci-ln-w076dpt-d5d6b.origin-ci-int-aws.dev.rhcloud.com,etcd-0.,etcd-1.,etcd-2.,localhost"
```

```
$ oc logs -n openshift-machine-config-operator machine-config-daemon-nhhsz -c machine-config-daemon
I0105 14:35:42.014625    5900 start.go:108] Version: machine-config-daemon-4.6.0-202006240615.p0-542-gc85d770f-dirty (c85d770f65130f48efa47ab3f309662877ad2d09)
I0105 14:35:42.018761    5900 start.go:121] Calling chroot("/rootfs")
I0105 14:35:42.018919    5900 rpm-ostree.go:261] Running captured: rpm-ostree status --json
I0105 14:35:42.284149    5900 daemon.go:224] Booted osImageURL: registry.svc.ci.openshift.org/ci-ln-w076dpt/stable@sha256:52c20c647d964d98066603a00412a18b876d6da1f7235e409f184794790f47a6 (47.83.202101041743-0)
I0105 14:35:42.404616    5900 daemon.go:231] Installed Ignition binary version: 2.8.1
I0105 14:35:42.497463    5900 start.go:97] Copied self to /run/bin/machine-config-daemon on host
I0105 14:35:42.500829    5900 update.go:1844] Starting to manage node: ip-10-0-220-241.us-west-1.compute.internal
I0105 14:35:42.500913    5900 metrics.go:105] Registering Prometheus metrics
I0105 14:35:42.502061    5900 metrics.go:110] Starting metrics listener on 127.0.0.1:8797
I0105 14:35:42.508337    5900 rpm-ostree.go:261] Running captured: rpm-ostree status
I0105 14:35:42.545589    5900 daemon.go:863] State: idle
Deployments:
* pivot://registry.svc.ci.openshift.org/ci-ln-w076dpt/stable@sha256:52c20c647d964d98066603a00412a18b876d6da1f7235e409f184794790f47a6
              CustomOrigin: Managed by machine-config-operator
                   Version: 47.83.202101041743-0 (2021-01-04T17:46:18Z)

  ostree://950ea2dee9038fd09c88883ad258b4c3fa91cde7155409fb385e348387074d46
                   Version: 47.83.202012030221-0 (2020-12-03T02:25:01Z)
I0105 14:35:42.545609    5900 rpm-ostree.go:261] Running captured: journalctl --list-boots
I0105 14:35:42.550667    5900 daemon.go:870] journalctl --list-boots:
-1 097e9a0e68984f7fa9557144f0aba666 Tue 2021-01-05 14:15:17 UTC—Tue 2021-01-05 14:32:00 UTC
 0 d5f8226b9f634475bd25823731ed1c6a Tue 2021-01-05 14:32:17 UTC—Tue 2021-01-05 14:35:42 UTC
I0105 14:35:42.550682    5900 rpm-ostree.go:261] Running captured: systemctl list-units --state=failed --no-legend
I0105 14:35:42.557422    5900 daemon.go:885] systemd service state: OK
I0105 14:35:42.557437    5900 daemon.go:617] Starting MachineConfigDaemon
I0105 14:35:42.557516    5900 daemon.go:624] Enabling Kubelet Healthz Monitor
I0105 14:35:43.514422    5900 daemon.go:395] Node ip-10-0-220-241.us-west-1.compute.internal is part of the control plane
I0105 14:35:45.622219    5900 node.go:24] No machineconfiguration.openshift.io/currentConfig annotation on node ip-10-0-220-241.us-west-1.compute.internal: map[volumes.kubernetes.io/controller-managed-attach-detach:true], in cluster bootstrap, loading initial node annotation from /etc/machine-config-daemon/node-annotations.json
I0105 14:35:45.622899    5900 node.go:45] Setting initial node config: rendered-master-d7e09bd37c2c9ec80abe64782130d660
I0105 14:35:45.665914    5900 daemon.go:781] In bootstrap mode
E0105 14:35:45.666005    5900 writer.go:135] Marking Degraded due to: machineconfig.machineconfiguration.openshift.io "rendered-master-d7e09bd37c2c9ec80abe64782130d660" not found
I0105 14:35:47.669738    5900 daemon.go:781] In bootstrap mode
E0105 14:35:47.670189    5900 writer.go:135] Marking Degraded due to: machineconfig.machineconfiguration.openshift.io "rendered-master-d7e09bd37c2c9ec80abe64782130d660" not found
I0105 14:36:19.689761    5900 daemon.go:781] In bootstrap mode
I0105 14:36:19.689789    5900 daemon.go:809] Current+desired config: rendered-master-d7e09bd37c2c9ec80abe64782130d660
I0105 14:36:19.696682    5900 daemon.go:1061] No bootstrap pivot required; unlinking bootstrap node annotations
I0105 14:36:19.696734    5900 daemon.go:1099] Validating against pending config rendered-master-d7e09bd37c2c9ec80abe64782130d660
I0105 14:36:19.716080    5900 daemon.go:1110] Validated on-disk state
I0105 14:36:19.748807    5900 daemon.go:1165] Completing pending config rendered-master-d7e09bd37c2c9ec80abe64782130d660
I0105 14:36:19.748895    5900 update.go:1844] completed update for config rendered-master-d7e09bd37c2c9ec80abe64782130d660
I0105 14:36:19.760796    5900 daemon.go:1181] In desired config rendered-master-d7e09bd37c2c9ec80abe64782130d660
```

File: pkg/operator/sync.go
Comment (ironcladlou) [reply pkg/operator/sync.go:850]: (No change required here, just a suggestion)

Comment (hexfusion) [reply pkg/operator/sync.go:850]: Yeah I think your approach is more readable having proxy object input and output makes this short term fix easier to revert. I have updated a hybrid.

File: pkg/operator/sync.go
Comment (ironcladlou) [inline pkg/operator/sync.go:847]: The way filtered is built up and then incrementally assigned to the field of the input argument is pretty unusual to me, at first I thought I was seeing a bug. To remove the side effects entirely and make it even more clear, what do you think about moving the precondition check to the caller, like:

```go
func (optr *Operator) getGlobalConfig() (*configv1.Infrastructure, *configv1.Net) {
	// <omitted>
  
	if proxy != nil && proxy.Status != (configv1.ProxyStatus{}) {
		proxy.Status.NoProxy = removeEtcdNoProxyConfig(proxy.Status.NoProxy)
	}
	return infra, network, proxy, dns, nil
}

func removeEtcdNoProxyConfig(lines string) string {
	filtered := []string{}
	for _, entry := range strings.Split(lines, ",") {
		if !regexp.MustCompile(`^etcd-.*$`).MatchString(entry) {
			filtered = append(filtered, entry)
		}
	}
	return strings.Join(filtered, ",")
}
```

File: pkg/operator/sync.go
Comment (ironcladlou) [reply pkg/operator/sync.go:850]: Continued in https://github.com/openshift/machine-config-operator/pull/2315/files#r552186937, not sure why I didn't reply here 🤦‍♀️

**Comment by hexfusion** [comment]
/hold 
When I initially put up this PR the intention was to tolertate a transient condition in CI. But the more I think about this, during the upgrade to 4.7 we need to also ensure the etcd records are removed from existing configs. For this reason more thought is required for the solution.

**Comment by hexfusion** [comment]
/hold cancel

At this point I am not 100% sure what other direction we would go here. I think this PR will mitigate any sync issues with etcd records in the NoProxy config. It seems though that a doc should be added to the release outlining manual removal of the stale records themselves. It is hard to say what the cluster might still be doing with these old etcd A records if we remove them we could break something. I think it would be reasonable to tolerate them and ask the user to remove?

If someone has another solution to automating removal please let me know. 

cc @kikisdeliveryservice @staebler PTAL

**Comment by hexfusion** [comment]
Prometheus issue..

/test e2e-aws

**Comment by staebler** [comment]
> /hold cancel
> 
> At this point I am not 100% sure what other direction we would go here. I think this PR will mitigate any sync issues with etcd records in the NoProxy config. It seems though that a doc should be added to the release outlining manual removal of the stale records themselves. It is hard to say what the cluster might still be doing with these old etcd A records if we remove them we could break something. I think it would be reasonable to tolerate them and ask the user to remove?
> 
> If someone has another solution to automating removal please let me know.
> 
> cc @kikisdeliveryservice @staebler PTAL

I am probably missing something, but it does not seem necessary to have the user do any manually removal. My understanding is that the conflict in the config only occurs when the installer generates something different than the cluster-network-operator, but that the conflict is only a problem during initial install and not for subsequent operation of the cluster. Once the cluster-network-operator takes over, then the configuration generated by the installer is moot. These changes in this PR are only necessary until both the installer and cluster-network-operator changes have merged.

These changes look fine to me, but I am not an owner in this repo.

**Comment by hexfusion** [comment]
Thanks, Matthew, yeah that was my initial read and I agree lets tolerate the CI changes and then revert.

**Comment by kikisdeliveryservice** [comment]
I'd like a successful e1e-aws run before approving (and it's reqd to merge anyway) but we're waiting on the kubelet fix upstream for ci to unblock.

But just to understand we will merge this, then the two prs above, then revert this?

/retest

**Comment by hexfusion** [comment]
> I'd like a successful e2e-aws run before approving (and it's reqd to merge anyway) but we're waiting on the kubelet fix upstream for ci to unblock.
> 
> But just to understand we will merge this, then the two prs above, then revert this?
> 
> /retest

correct

**Comment by hexfusion** [comment]
cc @kikisdeliveryservice think CI is back !

**Review by kikisdeliveryservice — APPROVED** [review]

File: pkg/operator/sync_test.go
Comment (marun) [inline pkg/operator/sync_test.go:29]: If the etcd address(es) were last in the line and were removed by `removeEtcdNoProxyConfig` that would result in removal of the newline. Is that possible, and is the trailing newline relevant?

File: pkg/operator/sync.go
Comment (marun) [inline pkg/operator/sync.go:840]: nit: s/temporary remove of/temporarily remove/

File: pkg/operator/sync_test.go
Comment (kikisdeliveryservice) [reply pkg/operator/sync_test.go:29]: oh good q!

**Comment by kikisdeliveryservice** [comment]
good q above from @marun 

/hold

File: pkg/operator/sync_test.go
Comment (staebler) [reply pkg/operator/sync_test.go:29]: I don't think that the lines will have that extra newline at the end.

File: pkg/operator/sync.go
Comment (hexfusion) [reply pkg/operator/sync.go]: done

File: pkg/operator/sync_test.go
Comment (hexfusion) [reply pkg/operator/sync_test.go:34]: I added support in case I don't want to break anything so if it has a newline it will persist. Its not beautiful but alas temporary.

**Comment by hexfusion** [comment]
Based on comments from @staebler speaking with @marun as well as reviewing API, we feel proxy should never contain the newline rune at end of the config. I have updated the tests to reflect.

https://github.com/openshift/machine-config-operator/blob/master/vendor/github.com/openshift/api/config/v1/types_proxy.go#L82

**Comment by marun** [comment]
LGTM

**Pull request merged**
